### PR TITLE
Relative and absolute equality

### PR DESCRIPTION
**Pull request by tmilnthorp**
Initial go at #391

### CODE REVIEW

File: UnitsNet/UnitDefinitions/Energy.json [hunk 1]
```diff
@@ -136,8 +136,8 @@
     {
       "SingularName": "ThermImperial",
       "PluralName": "ThermsImperial",
-      "FromUnitToBaseFunc": "x*1.05505585257348e+14",
-      "FromBaseToUnitFunc": "x/1.05505585257348e+14",
+      "FromUnitToBaseFunc": "x*1.05967e8",
```
Comment (tmilnthorp) [inline UnitsNet/UnitDefinitions/Energy.json]: [Was wrong](http://www.wolframalpha.com/input/?i=1+therm_39+to+J)

Comment (angularsen) [reply UnitsNet/UnitDefinitions/Energy.json]: There are actually 4 definitions: Therm EC/US/UK and BTU

https://en.wikipedia.org/wiki/Therm#Definitions

ThermImperial may be misleading in name, but refers to Therm UK.
There is already a unit for BritishThermalUnit (BTU) further up here.

So I think it is already correct, as is. Do you agree?

File: UnitsNet.Tests/AssertEx.cs [hunk 2]
```diff
@@ -7,9 +8,22 @@ namespace UnitsNet.Tests
...
         {
-            Assert.True(actual >= expected - tolerance && actual <= expected + tolerance, $"Values are not equal within tolerance: {tolerance}\nExpected: {expected}\nActual: {actual}\nDiff: {actual - expected:e}");
+            if(comparisonType == ComparisonType.Relative)
+            {
+                bool areEqual = UnitsNet.Comparison.EqualsRelative(expected, actual, tolerance);
+
+                double difference = Math.Abs(expected - actual);
+                double percentDifference = (difference / actual) * 100.0d;
+
+                Assert.True( areEqual, $"Values are not equal within relative tolerance: {tolerance * 100.0d}%\nExpected: {expected}\nActual: {actual}\nDiff: {percentDifference}%" );
```
Comment (angularsen) [inline UnitsNet.Tests/AssertEx.cs]: You can use format P2 as in `{0.0354321:P2}` to get `"3.54%"` string including the `%` symbol. It also uses the right symbol for the current culture.

Comment (angularsen) [reply UnitsNet.Tests/AssertEx.cs]: I would perhaps argue that percentage should be relative to `expected` and not `actual`, since you don't really know what actual is, but expected is .. well.. what you expect :-) If expected is `10` and actual is `0.000001`, then a difference of `~10` becomes a huge percentage instead of having an error of `~100%`.

File: UnitsNet/Comparison.cs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+using System;
+
+namespace UnitsNet
+{
+    public class Comparison
+    {
+        public static bool Equals( double value1, double value2, double tolerance = 0.00001, ComparisonType comparisonType = ComparisonType.Relative)
+        {
+            if(comparisonType == ComparisonType.Relative)
```
Comment (angularsen) [inline UnitsNet/Comparison.cs]: A `switch` can be slightly more readable when comparing the same value multiple times and returning a value based on it.

File: UnitsNet/Comparison.cs [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+                return EqualsRelative(value1, value2, tolerance);
+            else if(comparisonType == ComparisonType.Absolute)
+                return EqualsAbsolute(value1, value2, tolerance);
+            else
+                throw new InvalidOperationException("The ComparisonType is unsupported.");
+        }
+
+        public static bool EqualsRelative( double value1, double value2, double tolerance = 0.00001)
```
Comment (angularsen) [inline UnitsNet/Comparison.cs]: The xmldoc should explain that the tolerance is relative to `value1`, which is kind of important I think.
Consider renaming `value1` to `referenceValue`, since we must choose a reference to compare relative to. The other could be named `value` maybe. Perhaps there are some naming conventions on this sort of thing.

File: UnitsNet/Comparison.cs [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+        {
+            double maxVariation = Math.Abs( value1 * tolerance );
```
Comment (angularsen) [inline UnitsNet/Comparison.cs]: We should throw if tolerance is `< 0`, even if it would work the same.

File: UnitsNet/GeneratedCode/Extensions/Number/NumberToPowerDensityExtensions.g.cs [hunk 6]
```diff
@@ -1132,6 +1166,40 @@ public static class NumberToPowerDensityExtensions
 
         #endregion
 
+        #region PetawattPerCubicMeter
+
+        /// <inheritdoc cref="PowerDensity.FromPetawattsPerCubicMeter(UnitsNet.QuantityValue)" />
+        public static PowerDensity PetawattsPerCubicMeter(this int value) => PowerDensity.FromPetawattsPerCubicMeter(value);
```
Comment (angularsen) [inline UnitsNet/GeneratedCode/Extensions/Number/NumberToPowerDensityExtensions.g.cs]: Consider moving the PowerDensity units to a separate PR, even if that is probably why you started working on the abs vs relative comparison to begin with.

File: UnitsNet/GeneratedCode/Quantities/VolumeFlow.g.cs [hunk 7]
```diff
@@ -1153,6 +1153,25 @@ public override bool Equals(object obj)
...
+        ///     Relative tolerance is when the difference between the two quantities is not greater than the scale of the values compared to the tolerance.
+        ///     Absolute tolerance is when the difference between the two quantities is not greater than exactly the tolerance.
+        ///     Note that it is advised against specifying zero difference, due to the nature
+        ///     of floating point operations and using System.Double internally.
+        /// </summary>
+        /// <param name="other">The other quantity to compare to.</param>
+        /// <param name="tolerance">The comparison tolerance.</param>
+        /// <param name="comparisonType">The comparison type: either relative or absolute.</param>
+        /// <returns>True if the difference between the two values is not greater than the specified relative or absolute tolerance.</returns>
+        public bool Equals(VolumeFlow other, double tolerance = 0.00001, ComparisonType comparisonType = ComparisonType.Relative)
```
Comment (angularsen) [inline UnitsNet/GeneratedCode/Quantities/VolumeFlow.g.cs]: Tolerance probably needs to be of type `VolumeFlow`, since the consumer may be working with units of entirely different orders of magnitude than the base unit - so the default of `0.00001` may be way too small or way too large. 

A second option is to rely on the fact that we keep track of the `Unit` the quantity was constructed with, so we could pick one of the quantities' `Unit` and convert both quantities to a numeric value of that unit, then compare that with a default numeric tolerance. So when comparing a 1 nanometer with 2 nanometers, then the tolerance would 0.000001 nanometers by default rather than 0.000001 meters. The pitfall here is that it matters WHICH value you choose the unit for, as you get very different behavior when comparing 1 kilometer to 1e12 nanometers vs comparing 1e12 nanometers to 1 kilometer. In this particular example, I would expect `this` to be the reference - which I guess is intuitive.

File: UnitsNet/UnitDefinitions/Energy.json (hunk 1 above)
Comment (tmilnthorp) [reply UnitsNet/UnitDefinitions/Energy.json]: I did assume it was therm(UK). I thought that was therm_39, but looks like I'm completely wrong there. It caught my eye immediately though because the scale is completely off (e14).

If you look at Wikipedia, it should be 1.05505585257348e8, not 1.05505585257348e14. Looks like the decimal on Wikipedia got overlooked :)

So just to make sure I'm not missing anything, 1.05505585257348e8 seems correct, do you agree?

Comment (angularsen) [reply UnitsNet/UnitDefinitions/Energy.json]: Ah I missed that part :-D Yup, should be `e8`!

Comment (tmilnthorp) [reply UnitsNet/UnitDefinitions/Energy.json]: I will change it to that then!

File: UnitsNet.Tests/AssertEx.cs (hunk 2 above)
Comment (tmilnthorp) [reply UnitsNet.Tests/AssertEx.cs]: I didn't know you could do the P2 formatting outside of a ToString. Nice! I'll update that. And I agree that it should be against the expected.

File: UnitsNet/GeneratedCode/Extensions/Number/NumberToPowerDensityExtensions.g.cs (hunk 6 above)
Comment (tmilnthorp) [reply UnitsNet/GeneratedCode/Extensions/Number/NumberToPowerDensityExtensions.g.cs]: That is indeed why I started :) I just forgot to remove it. Good catch

File: UnitsNet/Comparison.cs (hunk 3 above)
Comment (tmilnthorp) [reply UnitsNet/Comparison.cs]: Done!

File: UnitsNet/Comparison.cs (hunk 4 above)
Comment (tmilnthorp) [reply UnitsNet/Comparison.cs]: I agree. I've updated this and added documentation that I hope is clear.

File: UnitsNet/Comparison.cs (hunk 5 above)
Comment (tmilnthorp) [reply UnitsNet/Comparison.cs]: Agreed. I was going to initially say <= 0 but I think it's perfectly valid to request being exact if desired.

File: UnitsNet/GeneratedCode/Quantities/VolumeFlow.g.cs (hunk 7 above)
Comment (tmilnthorp) [reply UnitsNet/GeneratedCode/Quantities/VolumeFlow.g.cs]: I think I like option #2 better. Absolute/relative can be a bit confusing and keeping it a double makes it easier to think of as a "percentage". Being typed would make sense for absolute differences, but since this is a generic method I think it's better as a double.

I made this in the same thinking of String.Equals where you can pass in the comparison type (ignore case, etc.).

I agree the reference being [this] is intuitive.

File: UnitsNet/Comparison.cs [hunk 8]
```diff
@@ -31,21 +31,21 @@ public sealed class Comparison
         ///     Absolute tolerance is when the absolute difference between the two values is not greater than the tolerance value.
         /// </summary>
         /// <param name="referenceValue">The first value. If using relative tolerance, it is the value which the relative tolerance will be calculated against.</param>
-        /// <param name="value">The value to compare to.</param>
+        /// <param name="otherValue">The value to compare to.</param>
         /// <param name="tolerance">The absolute or relative tolerance value.</param>
         /// <param name="comparisonType">Whether to use tolerance as an absolute or relative tolerance.</param>
         /// <returns></returns>
-        public static bool Equals(double referenceValue, double value, double tolerance, ComparisonType comparisonType)
+        public static bool Equals(double referenceValue, double otherValue, double tolerance, ComparisonType comparisonType)
```
Comment (angularsen) [inline UnitsNet/Comparison.cs]: WRC target is a big pain in the ass, I've contemplated dropping the support for it, but we have a Microsoft package that depends on it which is kind of cool.

I have created an issue to generate a separate code base for WRC, to avoid all the #if defs, which I think will be an improvement since most people don't care about WRC anyway. Then hopefully one will meet less friction when adding new stuff, since adding extra code like operator overloads would by default not be included in WRC.

https://github.com/angularsen/UnitsNet/issues/409

Comment (tmilnthorp) [reply UnitsNet/Comparison.cs]: I did see that, and considered picking that up :)

File: UnitsNet.Tests/CustomCode/MassTests.cs [hunk 9]
```diff
@@ -42,11 +42,11 @@ public class MassTests : MassTestsBase
 
         protected override double KilotonnesInOneKilogram => 1E-6;
 
-        protected override double LongTonsInOneKilogram => 0.000984207;
+        protected override double LongTonsInOneKilogram => 9.842065276110606e-4;
```
Comment (angularsen) [inline UnitsNet.Tests/CustomCode/MassTests.cs]: This PR is getting a bit crowded and not so easy to review. Could you perhaps pull out the bits not directly related to absolute/relative functionality in a separate PR?

Comment (tmilnthorp) [reply UnitsNet.Tests/CustomCode/MassTests.cs]: Unfortunately all of these changes are so the tests pass. These validation numbers were off quite a bit when I changed the tests to validate by % tolerance (not precise enough).

Comment (angularsen) [reply UnitsNet.Tests/CustomCode/MassTests.cs]: Haha 😆 it's so good to finally fix this!
I understand about the precision changes, no problem!

Comment (angularsen) [reply UnitsNet.Tests/CustomCode/MassTests.cs]: Is this getting ready now? I might find some time tomorrow to review.

Comment (tmilnthorp) [reply UnitsNet.Tests/CustomCode/MassTests.cs]: Yup! I'm done making changes.

File: UnitsNet/Comparison.cs [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+using System;
+
+namespace UnitsNet
+{
+    public sealed class Comparison
```
Comment (angularsen) [inline UnitsNet/Comparison.cs]: Make this class `static`? Should it be `internal`?

File: UnitsNet/GeneratedCode/Quantities/Acceleration.g.cs [hunk 11]
```diff
@@ -790,6 +790,29 @@ public override bool Equals(object obj)
             return AsBaseUnitMetersPerSecondSquared().Equals(((Acceleration) obj).AsBaseUnitMetersPerSecondSquared());
         }
 
+        /// <summary>
+        ///     Compare equality to another Acceleration within the given absolute or relative tolerance.
+        ///     Relative tolerance is when the difference between the two quantities is not greater than the scale of the values compared to the tolerance.
```
Comment (angularsen) [inline UnitsNet/GeneratedCode/Quantities/Acceleration.g.cs]: How about:
> Relative tolerance is defined by a fraction of this quantity, so a value of 0.01 means +/- 1% tolerance.
> Absolute tolerance is defined by the max difference between this quantity and <paramref name="other" /> in the <see cref="Unit" /> of this quantity. So if this quantity was constructed with 100 cm, then a tolerance of 0.01 means a tolerance of +/- 1 cm.

When writing this, I realize that I'm not sure this behavior for absolute tolerance is intuitive enough and whether we should have a separate method for absolute equality that instead take an `Acceleration` as tolerance? The relative tolerance implementation is fine I think.

File: UnitsNet/UnitDefinitions/Pressure.json [hunk 12]
```diff
@@ -272,8 +272,8 @@
     {
       "SingularName": "TonneForcePerSquareCentimeter",
       "PluralName": "TonnesForcePerSquareCentimeter",
-      "FromUnitToBaseFunc": "x*98066501.9960652",
-      "FromBaseToUnitFunc": "x*1.01971619222242E-08",
+      "FromUnitToBaseFunc": "x*9.80665e7",
+      "FromBaseToUnitFunc": "x/9.80665e7",
```
Comment (angularsen) [inline UnitsNet/UnitDefinitions/Pressure.json]: Nice that you are making this consistent with the guideline!

File: UnitsNet.Tests/CustomCode/MassTests.cs [hunk 13]
```diff
@@ -42,11 +42,11 @@ public class MassTests : MassTestsBase
...
 
         protected override double MegapoundsInOneKilogram => 2.2046226218487757e-6;
 
-        protected override double MegatonnesInOneKilogram => 1E-6;
+        protected override double MegatonnesInOneKilogram => 1E-9;
```
[the remark this answers is not in the document]
Comment (tmilnthorp) [reply UnitsNet.Tests/CustomCode/MassTests.cs]: I know!

File: UnitsNet/Comparison.cs (hunk 10 above)
Comment (tmilnthorp) [reply UnitsNet/Comparison.cs]: I thought I tried and Windows Runtime Component yelled at me. I'll double check because it should definitely be static. Not so sure about internal. Can be used anywhere if someone wishes!

File: UnitsNet/GeneratedCode/Quantities/Acceleration.g.cs (hunk 11 above)
Comment (tmilnthorp) [reply UnitsNet/GeneratedCode/Quantities/Acceleration.g.cs]: I think having an example is good here. Makes it easier to understand. I'll add that. 

As for making the method typed I'm not so sure. When we want to implement an IQuantity interface, having the generic method that takes a double is the only way to make it reusable as an interface method.

File: UnitsNet/UnitDefinitions/Pressure.json (hunk 12 above)
Comment (tmilnthorp) [reply UnitsNet/UnitDefinitions/Pressure.json]: When the tests failed due to relative checks, I couldn't know if the conversion was wrong or the test, so I did both :)

File: UnitsNet.Tests/CustomCode/DurationTests.cs [hunk 14]
```diff
@@ -36,19 +36,19 @@ public class DurationTests : DurationTestsBase
...
 
         protected override double NanosecondsInOneSecond => 1e+9;
 
         protected override double SecondsInOneSecond => 1;
 
-        protected override double WeeksInOneSecond => 1.6534e-6;
+        protected override double WeeksInOneSecond => 1.653439153439153e-6;
 
-        protected override double YearsInOneSecond => 3.1689e-8;
+        protected override double YearsInOneSecond => 3.170979198376458e-8;
```
[the remark this answers is not in the document]
Comment (tmilnthorp) [reply UnitsNet.Tests/CustomCode/DurationTests.cs]: Are you saying it is supposed to be different? Both are defines in the json files as x*365*24*3600, so I am not sure how they could be?

File: UnitsNet/Comparison.cs (hunk 10 above)
Comment (tmilnthorp) [reply UnitsNet/Comparison.cs]: It seems to be fine with static. I'll leave it as public though if that's ok with you?

File: UnitsNet/GeneratedCode/Quantities/Acceleration.g.cs (hunk 11 above)
Comment (tmilnthorp) [reply UnitsNet/GeneratedCode/Quantities/Acceleration.g.cs]: That is true with the interface, but generally I'd expect a check, much like Equals(Object) does. If the types are different it would just always return false. But all debatable :)

Are you saying you'd like to see:

```csharp
var a = Length.FromMeters( 5.0 );
var isEqual = a.EqualsAbsolute( Length.FromMeters( 10.0 ), Length.FromNauticalMiles( 100000 ) )
```

**Comment by tmilnthorp** [comment]
How on earth does public bool Equals($quantityName other, $quantityName maxError) in the source code generation not get emitted as a change in the generated files? :confused:

File: UnitsNet/Comparison.cs [hunk 15]
```diff
@@ -26,14 +26,23 @@ namespace UnitsNet
...
-        ///     Absolute tolerance is when the absolute difference between the two values is not greater than the tolerance value.
+        ///     </para>
+        ///     <para>
+        ///     Relative tolerance is defined as the maximum allowable absolute difference between <paramref name="referenceValue"/> and
+        ///     <paramref name="otherValue"/> as a percentage of <paramref name="referenceValue"/>. For example, a relative tolerance
+        ///     of 0.01 means the difference must be within +/- 1% of <paramref name="referenceValue"/> to be considered equal.
+        ///     </para>
+        ///     <para>
+        ///     Absolute tolerance is defined as the maximum allowable absolute difference between <paramref name="referenceValue"/> and
+        ///     <paramref name="otherValue"/> as a fixed number.
```
Comment (angularsen) [inline UnitsNet/Comparison.cs]: Fixed number, but in what unit?

File: UnitsNet/Comparison.cs [hunk 16]
```diff
@@ -52,12 +61,14 @@ public static bool Equals(double referenceValue, double otherValue, double toler
...
-        ///     Relative tolerance is when the difference between the two values is not greater than referenceValue * tolerance.
+        ///     Checks if two values are equal with a given relative tolerance.
+        ///     Relative tolerance is defined as the maximum allowable difference between <paramref name="referenceValue"/> and
+        ///     <paramref name="otherValue"/> as a percentage of <paramref name="referenceValue"/>. For example, a relative tolerance
+        ///     of 0.01 means the difference must be within +/- 1% of <paramref name="referenceValue"/> to be considered equal.
         /// </summary>
         /// <param name="referenceValue">The reference value which the tolerance will be calculated against.</param>
         /// <param name="otherValue">The value to compare to.</param>
-        /// <param name="tolerance"></param>
+        /// <param name="tolerance">The relative tolerance. Must be greater than 0.</param>
```
Comment (angularsen) [inline UnitsNet/Comparison.cs]: greater or equal

File: UnitsNet/Comparison.cs [hunk 17]
```diff
@@ -26,14 +26,23 @@ namespace UnitsNet
...
+        ///     Absolute tolerance is defined as the maximum allowable absolute difference between <paramref name="referenceValue"/> and
+        ///     <paramref name="otherValue"/> as a fixed number.
+        ///     </para>
         /// </summary>
-        /// <param name="referenceValue">The first value. If using relative tolerance, it is the value which the relative tolerance will be calculated against.</param>
+        /// <param name="referenceValue">The reference value. If using relative tolerance, it is the value which the relative tolerance will be calculated against.</param>
         /// <param name="otherValue">The value to compare to.</param>
-        /// <param name="tolerance">The absolute or relative tolerance value.</param>
-        /// <param name="comparisonType">Whether to use tolerance as an absolute or relative tolerance.</param>
+        /// <param name="tolerance">The absolute or relative tolerance value. Must be greater than 0.</param>
```
Comment (angularsen) [inline UnitsNet/Comparison.cs]: greater or equal

File: UnitsNet/Comparison.cs [hunk 18]
```diff
@@ -70,7 +81,8 @@ public static bool EqualsRelative(double referenceValue, double otherValue, doub
 
         /// <summary>
         ///     Checks if two values are equal with a given absolute tolerance.
-        ///     Absolute tolerance is when the absolute difference between the two values is not greater than the tolerance value.
+        ///     Absolute tolerance is defined as the maximum allowable absolute difference between <paramref name="referenceValue"/> and
+        ///     <paramref name="otherValue"/> as a fixed number.
```
Comment (angularsen) [inline UnitsNet/Comparison.cs]: Same, what unit?

File: UnitsNet/Scripts/Include-GenerateQuantitySourceCode.ps1 [hunk 19]
```diff
@@ -445,15 +445,25 @@ namespace UnitsNet
...
+        ///     <paramref name="other"/> as a fixed number.
+        ///     </para>
+        ///     <para>
         ///     Note that it is advised against specifying zero difference, due to the nature
         ///     of floating point operations and using System.Double internally.
+        ///     </para>
         /// </summary>
         /// <param name="other">The other quantity to compare to.</param>
-        /// <param name="tolerance">The comparison tolerance.</param>
+        /// <param name="tolerance">The absolute or relative tolerance value. Must be greater than 0.</param>
```
Comment (angularsen) [inline UnitsNet/Scripts/Include-GenerateQuantitySourceCode.ps1]: greater or equal

File: UnitsNet/Scripts/Include-GenerateQuantitySourceCode.ps1 [hunk 20]
```diff
@@ -445,15 +445,25 @@ namespace UnitsNet
...
-        ///     Absolute tolerance is when the difference between the two quantities is not greater than exactly the tolerance value.
+        ///     </para>
+        ///     <para>
+        ///     Relative tolerance is defined as the maximum allowable absolute difference between this quantity's value and
+        ///     <paramref name="other"/> as a percentage of this quantity's value. For example, a relative tolerance
+        ///     of 0.01 means the difference must be within +/- 1% of this quantity's value to be considered equal.
+        ///     </para>
+        ///     <para>
+        ///     Absolute tolerance is defined as the maximum allowable absolute difference between this quantity's value and
+        ///     <paramref name="other"/> as a fixed number.
```
Comment (angularsen) [inline UnitsNet/Scripts/Include-GenerateQuantitySourceCode.ps1]: Same, what unit?

File: UnitsNet/Comparison.cs (hunk 15 above)
Comment (tmilnthorp) [reply UnitsNet/Comparison.cs]: In this case it's just a double, so there is no unit. But when you're looking at the Equals on a quantity, the reference value is "this" so it's a fixed number of "this" unit.

```csharp
var a = Length.FromMeters( 100.0 );
var b = Length.FromInches( 5.0 );
a.Equals( b, 0.001 )
```

In this case it's absolute of a. So within 0.001 meters.

Comment (angularsen) [reply UnitsNet/Comparison.cs]: Okay sure, that is one way to look at it. I will still argue that there is a unit in play here and it is the constructed unit of `a`, which is the most sensible choice by all means! However, I would have to read the current xmldoc more than once to understand what will actually happen, since there is conversions going on here and you are naturally not comparing the values 100 and 5 head on. If the xmldoc rather said the tolerance was in the `referenceValue`'s `Unit` then at least in my eyes it would be much more intuitive. I guess we just have different perspective on this and read it a bit differently.

So for me, a slight change in wording of xmldoc would make it a lot more intuitive.

Comment (tmilnthorp) [reply UnitsNet/Comparison.cs]: I think changing the wording definitely makes sense regardless!

**Comment by tmilnthorp** [comment]
Updating the documentation on the Equals method in quantity classes

File: UnitsNet/Scripts/Include-GenerateQuantitySourceCode.ps1 [hunk 21]
```diff
@@ -444,6 +444,57 @@ namespace UnitsNet
...
+        ///     </code>
+        ///     </example>
+        ///     </para>
+        ///     <para>
+        ///     Note that it is advised against specifying zero difference, due to the nature
+        ///     of floating point operations and using System.Double internally.
+        ///     </para>
+        /// </summary>
+        /// <param name="other">The other quantity to compare to.</param>
+        /// <param name="tolerance">The absolute or relative tolerance value. Must be greater than 0.</param>
```
Comment (angularsen) [inline UnitsNet/Scripts/Include-GenerateQuantitySourceCode.ps1]: greater or equal to 0

Comment (angularsen) [reply UnitsNet/Scripts/Include-GenerateQuantitySourceCode.ps1]: Also update `Comparison` with the same documentation updates.

**Comment by tmilnthorp** [comment]
I will also obsolete Equals($quantityName other, $quantityName maxError) since we have a method that does relative and obsolete now. I'll also change the Equals(Object) implementation

**Comment by angularsen** [comment]
Awesome! Looks good, nice touch on fixing the old equals methods too.
Nuget 3.101 on the way out.
